packages/utils/telemetry-utils/src/telemetryEventBatcher.ts

```diff
@@ -0,0 +1,139 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { truncateToDecimalPlaces } from "./mathTools.js";
+import type {
+	ITelemetryGenericEventExt,
+	ITelemetryLoggerExt,
+	ITelemetryPerformanceEventExt,
+} from "./telemetryTypes.js";
+
+/**
+ * Expected type of the custom data passed into the logger.
+ */
+// type Data<TKey extends string> = { readonly [key in TKey]: number };
+
+/**
+ * Expected type of the custom data passed into the logger.
+ */
+interface IMeasuredCodeResult<TKey extends string> {
+	telemetryProperties?: { readonly [key in TKey]: number };
+}
+
+/**
+ * @remarks The logger is expected to be used for a single event type. If the set of `telmetryProperties` is different for different events, a separate `TelemetryEventBatcher` should be created for each event type.
+ * Telemetry class that accumulates user defined telemetry metrics {@link ICustomDataMap} and sends it to the {@link  ITelemetryLoggerExt} logger provided to this class every time the {@link TelemetryEventBatcher.log} the number of calls to this function reaches a number specified by {@link TelemetryEventBatcher.threshold}.
+ * @typeparam TMetrics - The set of keys that should be logged.
+ * E.g., `keyof Foo` for logging properties `bar` and `baz` from `type Foo = { bar: number, baz: number }`.
+ *
+ * @internal
+ */
+export class TelemetryEventBatcher<TMetrics extends string> {
+	/**
+	 * `codeDuration`: stores the average duration of the code passed into the logger.
+	 * `dataSums`: stores the sum of the custom data passed into the logger.
+	 * `dataMaxes`: stores the maximum value of the custom data passed into the logger.
+	 * `dataSums` and `dataMaxes` should share identical sets of properties.
+	 */
+	private codeDuration: number = 0;
+	private dataSums: { [key in TMetrics]?: number } = {};
+	private dataMaxes: { [key in TMetrics]?: number } = {};
+
+	// Counter to keep track of the number of times the log function is called.
+	private counter = 0;
+
+	public constructor(
+		/**
+		 * Custom properties to include in the telemetry performance event when it is written.
+		 */
+		private readonly eventBase: ITelemetryGenericEventExt,
+
+		/**
+		 * The logger to use to write the telemetry performance event.
+		 */
+		private readonly logger: ITelemetryLoggerExt,
+
+		/**
+		 * The number of logs to accumulate before sending the data to the logger.
+		 */
+		private readonly threshold: number,
+	) {}
+
+	/**
+	 * Executes the specified code and keeps track of execution time statistics.
+	 * @param codeToMeasure - The code to be executed and measured.
+	 * @param customData - Custom data to be logged.
+	 *
+	 * @returns Whatever the passed-in code block returns.
+	 */
+	public measure<T extends IMeasuredCodeResult<TMetrics>>(
+		codeToMeasure: () => T,
+		customData: Record<TMetrics, number>,
```

Comment:
I don't think we want the `customData` as something that gets passed by the caller, we should just use the `returnValue.telemetryProperties` from `codeToMeasure()`. If you think about the rebaser case, that's the only way we can do it. Otherwise the caller of `measure()` has to call the rebase helper function to get the measurements so they can pass them in, then give us that helper function to be run _again_ inside this method so we can measure its execution time, and that second execution could be at best unnecessary, at worst a bug.

Reply:
That's a good point, thanks for the correction. Changed to `this.accumulateAndLog(returnValue.telemetryProperties);`.